// express/routes/routes.js
var express = require("express");
var router  = express.Router();
var fs = require('fs');
var passport = require("passport");
var multer = require('multer');
var user = require("../models/user");
var snowflake = require('../middleware/snowflake.js');
var multipart = require('connect-multiparty');
var multipartMiddleware = multipart();
var controller = require('../models/controller');
var imgPath = "./uploads/";
var tweet = require("../models/tweet");
var Twit = require('twit');
var config = require('../middleware/twitter/config.js');

router.get("/",snowflake.isLoggedOut,(req, res)=>{
   res.render("homePage.ejs");
});

router.get("/signup", snowflake.isLoggedOut,(req,res)=>{
  res.render("signup.ejs");
});

router.get("/login",snowflake.isLoggedOut,(req,res)=>{
	res.render("login.ejs");
})

router.get("/contact", (req,res)=>{
  res.render("contact.ejs");
});

router.get("/mainPage",snowflake.isLoggedIn,(req,res)=>{
  res.render("mainPage.ejs");
});

router.get("/mainPage/refresh",snowflake.isLoggedIn,(req,res)=>{

		var T = new Twit(config);

		T.get('search/tweets', { result_type: 'recent',q: `from:${req.user.username}+-filter:retweets+-filter:replies`, count: 100 }, function(err, data, response) {
  			//console.log(data);

  			var tweetsId = [];

  			data.statuses.forEach(function(obj){

  					var newData = {
  					id : obj.id_str,
   					created_at : obj.created_at,
				    content : obj.text,
				    retweets : obj.retweet_count,
				    likes : obj.favorite_count
  				};

  				tweetsId.push(obj.id_str);

  				tweet.findOneAndUpdate({id: obj.id_str}, {$set: newData},{new: true, upsert: true, useFindAndModify: false});


  			});

  			//console.log(tweetsId);

  			//res.type('.js');
  			res.send(tweetsId);
		});

});


router.get("/mainPage/update",snowflake.isLoggedIn,(req,res)=>{

	tweet.find({id: req.query.tweetId}).then(doc => {
    console.log(doc);
  });

});


router.get("/settings",snowflake.isLoggedIn, (req,res)=>{
  res.render("settings.ejs");
});

router.get("/home", (req,res)=>{
  res.redirect("/");
});


//Authentication Routes

router.post("/registration", (req,res)=>{

	//PASSPORT AUTOMATICALLY HANDLES DUPLICATED USERNAMES!

	var newUser = new user({
    image:"",
		username: req.body.username,
		firstName: req.body.firstName,
		lastName: req.body.lastName,
		email: req.body.email,
		gender: req.body.gender,
		accType: req.body.accType
	});
	user.register(newUser , req.body.password , (err,user)=>{

		if(err){
			console.log(err);
			req.flash("error", err.message);
			return res.redirect("/signup");
		}

		passport.authenticate("local")(req,res, ()=>{

			res.redirect("/mainPage");
		});
	});
});


router.post("/login", passport.authenticate("local",
{
	successRedirect:"/mainPage",
	failureRedirect:"/login",
	failureFlash: 'Invalid username or password.'
}), (req,res)=>{

});

router.get("/logout", (req,res)=>{
  req.logout();
  res.redirect("/");
});

router.post("/changepicture",multipartMiddleware,controller.uploadimage,(req,res)=>{
  res.redirect("/settings");
});

router.post("/changeemail",multipartMiddleware,controller.updatingemail,(req,res)=>{
  res.redirect("/settings");
});

router.post("/changePassword",multipartMiddleware,controller.updatingpassword,(req,res)=>{
  res.redirect("/settings");
});

module.exports = router;
