// express/routes/routes.js
var express = require("express");
var router  = express.Router();
var passport = require("passport");
var user = require("../models/user");
var snowflake = require('../middleware/snowflake.js');


router.get("/",snowflake.isLoggedOut,(req, res)=>{
   res.render("homePage.ejs");
});

router.get("/signup", snowflake.isLoggedOut,(req,res)=>{
  res.render("signup.ejs");
});

router.get("/login",snowflake.isLoggedOut,(req,res)=>{
	res.render("login.ejs");
})

router.get("/contact", (req,res)=>{
  res.render("contact.ejs");
});

router.get("/mainPage",snowflake.isLoggedIn,(req,res)=>{
  res.render("mainPage.ejs");
});

router.get("/home", (req,res)=>{
  res.redirect("/");
});

//Authentication Routes

router.post("/registration", (req,res)=>{

	//PASSPORT AUTOMATICALLY HANDLES DUPLICATED USERNAMES!

	var newUser = new user({
		username: req.body.username,
		firstName: req.body.firstName,
		lastName: req.body.lastName,
		email: req.body.email,
		gender: req.body.gender,
		accType: req.body.accType
	});
	user.register(newUser , req.body.password , (err,user)=>{

		if(err){
			console.log(err);
			req.flash("error", err.message);
			return res.redirect("/signup");
		}

		passport.authenticate("local")(req,res, ()=>{

			res.redirect("/mainPage");
		});
	});
});

router.post("/login", passport.authenticate("local",
{
	successRedirect:"/mainPage",
	failureRedirect:"/login",
	failureFlash: 'Invalid username or password.'
}), (req,res)=>{

});

router.get("/logout", (req,res)=>{
  req.logout();
  res.redirect("/");
});


module.exports = router;
